ibm_was/datadog_checks/ibm_was/ibm_was.py

```diff
@@ -0,0 +1,127 @@
+            'CountStatistic': self.monotonic_count,
+            'DoubleStatistic': self.rate,
+            'RangeStatistic': self.gauge,
+            'TimeStatistic': self.gauge
+        }
+
+    def check(self, instance):
+        validation.validate_config(instance)
+        collect_stats = self.setup_configured_stats(instance)
+        url = instance.get('servlet_url')
+
+        nested_tags, metric_categories = self.append_custom_queries(instance, collect_stats)
+        custom_tags = instance.get('tags', [])
+
+        service_check_tags = list(custom_tags)
+        service_check_tags.append('url:{}'.format(url))
+
+        data = self.make_request(instance, url, service_check_tags)
+        server_data_xml = etree.fromstring(data)
+        node_list = self.get_node_from_root(server_data_xml, "Node")
+
+        for node in node_list:
+            server_list = self.get_node_from_root(node, 'Server')
+            node_tags = list(custom_tags)
+
+            node_tags.append('node:{}'.format(node.get('name')))
+            for server in server_list:
+                server_tags = ['server:{}'.format(server.get('name'))]
+                server_tags.extend(node_tags)
+
+                for category, prefix in iteritems(metric_categories):
+                    self.log.debug("Collecting {} stats".format(category))
+                    if collect_stats.get(category):
+                        stats = self.get_node_from_name(server, category)
+                        self.process_stats(stats, prefix, metric_categories, nested_tags, server_tags)
+
+    def get_node_from_name(self, xml_data, path):
+        # XMLPath returns a list, but there should only be one element here since the function starts
+        # the search within a given Node/Server
+        data = xml_data.xpath('//Stat[normalize-space(@name)="{}"]'.format(path))
+        if len(data):
+            return data[0]
+        else:
+            self.warning('Error finding {} stats in XML output.'.format(path))
+            return []
+
+    def get_node_from_root(self, xml_data, path):
+        return xml_data.xpath('//{}'.format(path))
+
+    # The XML will have Stat Nodes and Nodes that contain the metrics themselves
+    # This code recursively goes through each Stat Node to properly setup tags
+    # where each Stat will have a different tag key depending on the context.
+    def process_stats(self, stats, prefix, metric_categories, nested_tags, tags, recursion_level=0):
+        for child in stats:
+            if child.tag in metrics.METRIC_VALUE_FIELDS:
+                self.submit_metrics(child, prefix, tags)
+            elif child.tag in metrics.CATEGORY_FIELDS:
+                recursion_tags = tags + ["{}:{}".format(
+                    nested_tags.get(prefix)[recursion_level], child.get('name')
+                )]
+                self.process_stats(child, prefix, metric_categories, nested_tags, recursion_tags, recursion_level+1)
```

Comment:
Python actually has a recursion limit. How deep can these xml files go? (It might well not be deep enough to hit that limit, it's pretty high but not infinite)

Reply:
This should be pretty small. From the data I've seen, maybe 3 or 4 max so far.

Reply:
then we're fine